File: 739. Daily Temperatures/solution.py
class Solution:
    def dailyTemperatures(self, T):
        """
        :type T: List[int]
        :rtype: List[int]
        """
        sol = []
        stack = []
        for i in range(len(T)-1, -1, -1):
            while stack and T[stack[-1]] <= T[i]:
                stack.pop()
            
            if not stack:
                sol.append(0)
            else:
                sol.append(stack[-1] - i)
                
            stack.append(i)
            
        return sol[::-1]

File: 739. Daily Temperatures/solution.java
class Solution {
    public int[] dailyTemperatures(int[] T) {
        Stack<Integer> stack = new Stack<Integer>();
        int n = T.length;
        int[] sol = new int[n];

        n--;
        while (n >= 0) {
            while (!stack.isEmpty() && T[stack.peek()] <= T[n]) {
                stack.pop();
            }
            
            if (!stack.isEmpty())
                sol[n] = stack.peek() - n;
            else 
                sol[n] = 0;
            
            stack.add(n);
            n--;
        }
        return sol;
    }
}
